program/tests/interface.rs

```diff
@@ -0,0 +1,830 @@
+                        destination_status,
+                        true,
+                        lockup_state.to_lockup(CUSTODIAN_LEFT),
+                    ),
+                    minimum_delegation,
+                );
+
+                instruction::move_lamports(
+                    &STAKE_ACCOUNT_BLACK,
+                    &STAKE_ACCOUNT_WHITE,
+                    &STAKER_GRAY,
+                    free_lamports,
+                )
+            }
+            Self::Withdraw(source_has_delegation, lockup_state, full_withdraw) => {
+                let free_lamports = LAMPORTS_PER_SOL;
+                let source_status = match source_has_delegation {
+                    None => StakeStatus::Uninitialized,
+                    Some(false) => StakeStatus::Initialized,
+                    Some(true) => StakeStatus::Active,
+                };
+
+                env.update_stake(
+                    &STAKE_ACCOUNT_BLACK,
+                    &fully_configurable_stake(
+                        VOTE_ACCOUNT_RED,
+                        STAKE_ACCOUNT_BLACK,
+                        minimum_delegation,
+                        source_status,
+                        false,
+                        lockup_state.to_lockup(CUSTODIAN_LEFT),
+                    ),
+                    minimum_delegation + free_lamports,
+                );
+
+                let withdraw_amount = if full_withdraw && source_status != StakeStatus::Active {
+                    free_lamports + minimum_delegation + STAKE_RENT_EXEMPTION
+                } else {
+                    free_lamports
+                };
+
+                let authority = if source_status == StakeStatus::Uninitialized {
+                    STAKE_ACCOUNT_BLACK
+                } else {
+                    WITHDRAWER_BLACK
+                };
+
+                instruction::withdraw(
+                    &STAKE_ACCOUNT_BLACK,
+                    &authority,
+                    &PAYER,
+                    withdraw_amount,
+                    lockup_state.to_custodian(&CUSTODIAN_LEFT),
+                )
+            }
+            Self::Deactivate(lockup_state) => {
+                env.update_stake(
+                    &STAKE_ACCOUNT_BLACK,
+                    &fully_configurable_stake(
+                        VOTE_ACCOUNT_RED,
+                        STAKE_ACCOUNT_BLACK,
+                        minimum_delegation,
+                        StakeStatus::Active,
+                        false,
+                        lockup_state.to_lockup(CUSTODIAN_LEFT),
+                    ),
+                    minimum_delegation,
+                );
+
+                instruction::deactivate_stake(&STAKE_ACCOUNT_BLACK, &STAKER_BLACK)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Arbitrary)]
+enum StakeStatus {
+    Uninitialized,
+    Initialized,
+    Activating,
+    Active,
+    Deactivating,
+    Deactive,
+}
```

Comment:
Doesn't need to be done in this PR, but we could consider adding the `Arbitrary` derivation on the types within the interface crate with a `dev-context-only-utils` feature

Reply:
im not sure what i would use them for, there are so many external data dependencies (authorities on concrete keypairs, lockups on keypairs and clock, activation/deactivation on clock, delegation and rent exempt reserve on account lamports), invalid states, and valid states that are astronomically unlikely for a randomly generated struct to hit. i wrote my own so that `Arbitrary` would cover all important valid, and no invalid, states

Reply:
Definitely, I don't mean to do it for all of them, for the exact reason you outlined. But wherever possible, it's nice to avoid type duplication